x/sequencer/keeper/rotation.go

```diff
@@ -0,0 +1,133 @@
+// A process of rotation is defined by the time between when the proposer notice period is over till the proposer sends his last batch.
+func (k Keeper) IsRotating(ctx sdk.Context, rollappId string) bool {
+	return k.isNextProposerSet(ctx, rollappId)
+}
+
+// isNoticePeriodRequired returns true if the sequencer requires a notice period before unbonding
+// Both the proposer and the next proposer require a notice period
+func (k Keeper) isNoticePeriodRequired(ctx sdk.Context, seq types.Sequencer) bool {
+	return k.isProposer(ctx, seq.RollappId, seq.Address) || k.isNextProposer(ctx, seq.RollappId, seq.Address)
+}
+
+// ExpectedNextProposer returns the next proposer for a rollapp
+func (k Keeper) ExpectedNextProposer(ctx sdk.Context, rollappId string) types.Sequencer {
+	// if nextProposer is set, were in the middle of rotation
+	seqAddr, ok := k.GetNextProposerAddr(ctx, rollappId)
+	if ok {
+		return k.MustGetSequencer(ctx, seqAddr)
+	}
+
+	seqs := k.GetSequencersByRollappByStatus(ctx, rollappId, types.Bonded)
+	if len(seqs) == 0 {
+		return types.Sequencer{}
+	}
+
+	// take the next bonded sequencer to be the proposer. sorted by bond
+	sort.SliceStable(seqs, func(i, j int) bool {
+		return seqs[i].Tokens.IsAllGT(seqs[j].Tokens)
+	})
+
+	// return the first sequencer that is not the proposer
+	proposer, _ := k.GetProposer(ctx, rollappId)
+	for _, s := range seqs {
+		if s.Address != proposer.Address {
+			return s
+		}
+	}
+
+	return types.Sequencer{}
+}
+
+// startRotation sets the nextSequencer for the rollapp.
+// This function will not clear the current proposer
+// This function called when the sequencer has finished its notice period
+func (k Keeper) startRotation(ctx sdk.Context, rollappId string) {
+	// next proposer can be empty if there are no bonded sequencers available
+	nextProposer := k.ExpectedNextProposer(ctx, rollappId)
+	k.setNextProposer(ctx, rollappId, nextProposer.Address)
+
+	k.Logger(ctx).Info("rotation started", "rollappId", rollappId, "nextProposer", nextProposer.Address)
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeRotationStarted,
+			sdk.NewAttribute(types.AttributeKeyRollappId, rollappId),
+			sdk.NewAttribute(types.AttributeKeyNextProposer, nextProposer.Address),
+		),
+	)
+}
+
+// RotateProposer completes the sequencer rotation flow.
+// It's called when a last state update is received from the active, rotating sequencer.
+// it will start unbonding the current proposer, and set new proposer from the bonded sequencers
+func (k Keeper) RotateProposer(ctx sdk.Context, rollappId string) {
+	nextProposerAddr, ok := k.GetNextProposerAddr(ctx, rollappId)
+	if !ok { // nextProposer is guaranteed to be set by caller
+		k.Logger(ctx).Error("next proposer not set. rotation didn't completed", "rollappId", rollappId)
+		return
+	}
+
+	// start unbonding the current proposer
+	proposer, ok := k.GetProposer(ctx, rollappId)
+	if ok {
+		k.startUnbondingPeriodForSequencer(ctx, &proposer)
+	}
```

Comment:
can move this block above k.GetNextProposerAddr block

Reply:
why?